File: practica parte 1/Ejercicio 3: Clase genérica Catalogo/catalogo.py
class Catalogo:
    def __init__(self):
        self.elementos = []
    
    def agregar(self, item):
        self.elementos.append(item)
    
    def buscar(self, criterio):
        for item in self.elementos:
            if criterio.lower() in str(item).lower():
                return item
        return None
    
    def mostrar_todos(self):
        for item in self.elementos:
            print(item)

# Prueba con libros
class Libro:
    def __init__(self, titulo, autor):
        self.titulo = titulo
        self.autor = autor
    
    def __str__(self):
        return f"Libro: {self.titulo} - {self.autor}"

catalogo_libros = Catalogo()
catalogo_libros.agregar(Libro("Python 101", "Juan Perez"))
catalogo_libros.agregar(Libro("Java Básico", "Maria Gomez"))

print("\nBuscando libro 'Python':")
print(catalogo_libros.buscar("Python"))
print("\nTodos los libros:")
catalogo_libros.mostrar_todos()

File: practica parte 1/Ejercicio 3: Clase genérica Catalogo/src/catalogo.java
import java.util.ArrayList;
import java.util.List;

class Catalogo<T> {
    private List<T> elementos;
    
    public Catalogo() {
        elementos = new ArrayList<>();
    }
    
    public void agregar(T item) {
        elementos.add(item);
    }
    
    public T buscar(String criterio) {
        for (T item : elementos) {
            if (item.toString().toLowerCase().contains(criterio.toLowerCase())) {
                return item;
            }
        }
        return null;
    }
    
    public void mostrarTodos() {
        for (T item : elementos) {
            System.out.println(item);
        }
    }
}

class Libro {
    private String titulo;
    private String autor;
    
    public Libro(String titulo, String autor) {
        this.titulo = titulo;
        this.autor = autor;
    }
    
    @Override
    public String toString() {
        return "Libro: " + titulo + " - " + autor;
    }
}

public class MainCatalogo {
    public static void main(String[] args) {
        Catalogo<Libro> catalogoLibros = new Catalogo<>();
        catalogoLibros.agregar(new Libro("Python 101", "Juan Perez"));
        catalogoLibros.agregar(new Libro("Java Básico", "Maria Gomez"));
        
        System.out.println("\nBuscando libro 'Python':");
        System.out.println(catalogoLibros.buscar("Python"));
        
        System.out.println("\nTodos los libros:");
        catalogoLibros.mostrarTodos();
    }
}